pandas/_libs/properties.pyi

```diff
@@ -1,9 +1,28 @@
-# pyright: reportIncompleteStub = false
-from typing import Any
+from typing import (
+    Sequence,
+    overload,
+)
+
+from pandas._typing import (
+    AnyArrayLike,
+    DataFrame,
+    Index,
+    Series,
+)
 
 # note: this is a lie to make type checkers happy (they special
 # case property). cache_readonly uses attribute names similar to
 # property (fget) but it does not provide fset and fdel.
 cache_readonly = property
 
-def __getattr__(name: str) -> Any: ...  # incomplete
+class AxisProperty:
+
+    axis: int
+    def __init__(self, axis: int = ..., doc: str = ...) -> None: ...
+    @overload
+    def __get__(self, obj: DataFrame | Series, type) -> Index: ...
+    @overload
+    def __get__(self, obj: None, type) -> AxisProperty: ...
+    def __set__(
+        self, obj: DataFrame | Series, value: AnyArrayLike | Sequence
```

Comment:
type annotation for `value` chosen just to match `_set_axis` which is added to match the existing `ensure_index` annotation.

Reply:
Just an idea, might be not worth it to enforce that `obj` is compatible with `value`.

```py
IndexCompatibleT = TypeVar("IndexCompatibleT", AnyArrayLike | Sequence)

class Has_set_axis(Protocol[IndexCompatibleT]):
    def _set_axis(self: NDFrame, axis: int, value: IndexCompatibleT) -> object:
        ...


def __set__(self, obj: Has_set_axis[IndexCompatibleT], value: IndexCompatibleT) -> None:
    ...
```

Reply:
the index/column properties are only actually added to DataFrame/Series in the codebase.

In theory they could be added to any NDFrame object and I suppose we could refactor the code to add the AxisProperty dynamically in the base class based on the other NDFrame attributes, i.e. we know the axis numbers and names.

Having said that, i'm -1 on doing this dynamically and therefore happy with the DataFrame|Series to match the actual code.